chore: add vagrant support (#225)
arrocke/gloss-translation: docs/vagrant.md

   1. Launch the `Local Security Policy` app from the `start menu` (or `Win+R`, then type `secpol.msc`)
   1. Navigate to `Local Policies` -> `User Rights Assignment`.
   1. Open the `Create symbolic links` property. By default it has only `Administrators` listed.
   1. Click `Add User or Group...` button and type `Authenticated Users` in the `Enter the object names to select` field and then press the `OK` button. This will grant `symbolic link` priviledge to all users who log on with credetials.
   1. Close any open windows by clicking `OKs` and and then `log off` and `log on` to activate.
   1. Symbolic links should now work in VirtualBox shared folders without administrator priviledges.

## Setup Steps

1. [Download](https://developer.hashicorp.com/vagrant/downloads?ajs_aid=ba208f95-d5a1-457d-abeb-49d458b95eec&product_intent=vagrant) Vagrant software.
1. Reference these [installation](https://developer.hashicorp.com/vagrant/docs/installation) instructions.
1. Verify installation by running `vagrant -v` in a terminal session.
1. Install [VirtualBox](https://www.virtualbox.org/).
1. If you have not already done so, fork repo to your local machine.

> 1. Next you will need to install a [box](https://developer.hashicorp.com/vagrant/tutorials/getting-started/getting-started-boxes). The `Vagrantfile` in this repository uses box: `generic/ubuntu2204`. So to install this box, run in a terminal window:

Reviewer: If vagrant has already been initialized, then you don't need to install the box. Running `vagrant up` will install the box if it isn't already installed.

Author: I removed the install box section as you suggested.

```bash
    vagrant box add generic/ubuntu2204
```

To see which boxes are installed run:

```bash
    vagrant box list
```

> 7. Create a new folder and name it something that identifies the VM you are about to create. Example: `UbuntuVM`. The name of this folder will be used by Vagrant to identify this specific VM. It is to this directory that you will copy the vagrant file we have created for you. In a later step, Vagrant will create subdirectory`.vagrant` which it will use to manage your VM. It is in this directory where you will run all Vagrant commands that are specific to your VM.

Reviewer 2: @arrocke It almost sounds to me like we should check `UbuntuVM` (or another name) into the git repo, and not bother with having a top-level `.vagrant` directory that people have to copy from.

What do you think?

@ilearner777 Would my idea cause any issues?

Reviewer 2: I think if we did this, we could remove point 10 and just have `HOST_FOLDER = "../"` in the `Vagrantfile`

Author: I see there is a disconnect in our conversation.  This has to do with how I setup my directory structure when learning and developing this code with respect to what you are visualizing.  Let me explain...

In my directory structure, UbuntuVM is in a parallel directory to the repo.  This is why I have `HOST_FOLDER = .../"gloss-translation"` as the host URL. 
![image](https://github.com/arrocke/gloss-translation/assets/141659916/a0183f19-1dd0-4ffa-a0d3-d476b287ac6b)
I believe you are visualizing the UbuntuVM folder within the gloss-translation repository. 
Is this correct?

If so, I believe this is a better approach and found out that this is the preferred method.

So I am suggesting the following changes to this PR:

***Revised Directory Structure***
![image](https://github.com/arrocke/gloss-translation/assets/141659916/bc046e72-fca4-426f-8b3a-37e91867e303)

***Notes:***
- I have added a sub-directory `virualbox` to folder `.vagrant`. This is where I have placed the Vagrantfile configured for a Virtualbox VM. This is because I have already developed a Vagrantfile for the Docker environment. I plan to add a new issue/PR for this new environment. At that time I plan to add another sub-directory under `.vagrant` called `docker` into which I will place the Vagrantfile with configuration to create a docker VM. 
![image](https://github.com/arrocke/gloss-translation/assets/141659916/a3543707-474d-4251-baa1-b7f27b891d6e)
- I have renamed directory `UbuntuVM` to `vm` in order to keep it more generic, just in case someone wants an O/S other than Ubuntu.
- I have updated the .gitignore file to exclude the contents of the `vm` directory.  This is because when the user runs `Vagrant up`, files will be created that are specific to that user's machine.  These should not be stored in the upstream repository. See excerpt below:

    ```text
    # Virtual Machine
    vm/*
    !vm/.gitkeep
    ```

Please let me know if you approve of this recommendation and if you would like to modify anything I have described above.  I will await your reply before proceeding any further.

Reviewer: Personally, I think the Vagrantfile should be in the root directory and run from the root directory. As soon as we have to copy it, now there are going to be syncing issues if users haven't copied the latest version. We can add `.vagrant/` to `.gitignore` since that is produced at runtime.

8. Copy file: `Vagrantfile` from host `gloss-translation` repository `.vagrant/` folder to the new folder you created above.
9. Read the `NOTES:` section contained in file `Vagrantfile`.
10. in the `Vagrantfile`, edit `HOST_FOLDER` by replacing `../gloss-translation` with the location of your repository root folder.

```ruby
    HOST_FOLDER = "../gloss-translation"
```

> 11. If required, perform any requested `plugin` installations mentioned in the `NOTES:` section. To see which plugins are installed run:

Reviewer: The plugins that need to be installed should be documented here, not in the vagrant file

Author: Your recommendation to put Vagrantfile in the root directory is fine with me and I see the benefit of it.  However, I also see a downside in that then, we could only support a single provider. This is because we could only support a single version of the Vagrantfile.  

I have already created and successfully tested another version of the Vagrantfile that creates a VM using docker.  If you want to offer contributors the option to choose between multiple providers we would have to come up with a different approach. This was the reason why my design had the contributor copy the Vagrantfile of their choice to the VM folder.

I am good with whatever you wish to do.  Please let me know if you want to stick with offering only VirtualBox, or reconfigure with the  option to choose between multiple providers.

Reviewer: [This from the vagrant ](https://developer.hashicorp.com/vagrant/docs/providers/configuration)docs suggests that multiple providers is not an issue. These lines from the vagrantfile are ignored if you use a different provider:
```ruby
config.vm.provider "virtualbox" do |vb|
    vb.memory = 4096  # Set RAM in MB (4GB in this example)
    vb.cpus = 2      # Set the number of CPU cores
    vb.customize ["setextradata", :id, "VBoxInternal2/SharedFoldersEnableSymlinksCreate//home/vagrant/vmrepo", "1"]
  end
```

Everything else should be the same regardless of provider and anything that is required for only VirtualBox should be within that block

Author: Thanks for pointing that out to me.  Was not aware of this.  So I will update Vagrantfile so it is compatible with both VirtualBox and Docker. Then will implement all of your recommendations.  When complete will request another review.

```bash
    vagrant plugin list
```

> 12. To start the creation of your virtual development environment, run this command in your newly created directory:

Reviewer 2: It's not very clear, will we also run this to start the VM in the future, or is this command just for the initial creation of the virtual machine? How do we safely stop/shutdown the VM?

```bash
    vagrant up
```

> 13. For help on vagrant commands, at a terminal session enter:

Reviewer: We should mention that you need to run `vagrant ssh` to get access to the build environment to start the dev servers.

```bash
    vagrant -help
```

## Post-Installation Notes

Accessing VM resources from Host.

### Accessing database via pgAdmin

1. Start pgAdmin
1. Open the `Register - Server` dialog
1. `General` tab:
   - Name: _Enter name of your choice_
1. `Connection` tab

   - Host name: `127.0.0.1`
   - Port: `5432`
   - Maintenance database: `postgres`
   - Username: `vagrant`

> 1. `SSH Tunnel` tab

Reviewer 2: You forgot to mention the value for "Tunnel host".

Also, could we save the password in this tab to save time later?

   - Use SSH tunneling: `on`
   - Tunnel port: `2222`
   - Username: `vagrant`
   - Authentication: `Password`

### Accessing API app

Open a web brower to URL: <http://localhost:4300>

### Accessing Web app

Open a web brower to URL: <http://localhost:4200>

### Access VM via ssh

> - host Name: `localhost`
> - Port: `2222`
> - login name: `vagrant`
> - password: `vagrant`

Reviewer 2: ```suggestion
| Setting  | Value  |
| ------------- | ------------- |
|  hostname | `localhost`  |
| port  | 2222  |
| username  | `vagrant`  |
| passowrd  | `vagrant`  |
```

Reviewer 2: You could also use table formatting for the Connection tab item and SSH Tunnel tab settings, if you want.

Author: Done.

# Automated Ubuntu Local Env Setup

You can optionally setup a virtual Ubuntu environment using Vagrant. It is a tool used for building and managing virtual machine environments in a single workflow.

Although Vagrant supports [VirtualBox](https://www.virtualbox.org/), [Hyper-V](https://learn.microsoft.com/en-us/virtualization/hyper-v-on-windows/quick-start/enable-hyper-v), [Docker](https://www.docker.com/), or [VMware Fusion](https://customerconnect.vmware.com/downloads/get-download?downloadGroup=FUS-PUBTP-2021H1), `VirtualBox` is the only environment that has been developed and tested thus far. If you have need to use Hyper-V, Docker, or VMware, please create an issue and request a new feature.

## Prerequisites

1. You will require 4 GB of memory.

> 1. You will require 8 GB of diskspace.

Reviewer 2: ```suggestion
1. You will require 8 GB of disk space.
```

Author: Done

   1. Launch the `Local Security Policy` app from the `start menu` (or `Win+R`, then type `secpol.msc`)
   1. Navigate to `Local Policies` -> `User Rights Assignment`.
   1. Open the `Create symbolic links` property. By default it has only `Administrators` listed.
   1. Click `Add User or Group...` button and type `Authenticated Users` in the `Enter the object names to select` field and then press the `OK` button. This will grant `symbolic link` priviledge to all users who log on with credetials.
   1. Close any open windows by clicking `OKs` and and then `log off` and `log on` to activate.
   1. Symbolic links should now work in VirtualBox shared folders without administrator priviledges.

## Setup Steps

1. [Download](https://developer.hashicorp.com/vagrant/downloads?ajs_aid=ba208f95-d5a1-457d-abeb-49d458b95eec&product_intent=vagrant) Vagrant software.
1. Reference these [installation](https://developer.hashicorp.com/vagrant/docs/installation) instructions.
1. Verify installation by running `vagrant -v` in a terminal session.
1. Install [VirtualBox](https://www.virtualbox.org/).
1. If you have not already done so, fork repo to your local machine.

> 1. Create a new folder to contain your `Vagrant project`. Start a terminal session in that folder.

Reviewer 2: ```suggestion
1. Create a new folder to contain your Vagrant project. Start a terminal session in that folder.
```
It's not clear: should my vagrant project folder be inside my repo? Or somewhere else?

Author: I have completely rewritten this section, improving clarity and providing details as to the purpose of the new folder to be created.

1. Install a [box](https://developer.hashicorp.com/vagrant/tutorials/getting-started/getting-started-boxes). The `Vagrantfile` in this repository uses `generic/ubuntu2204`. To install it run:

> ```bash
> vagrant box add generic/ubuntu2204
> ```

Reviewer 2: ```suggestion
    ```bash
    vagrant box add generic/ubuntu2204
    ```
```
Tip: add indentation of four spaces before your code blocks to make them show up inline with your numbered list. (I've only commented on the one block of code, but this suggestion should be applied to all the code blocks within this numbered list)

Author: Done.

Author: Done.

Reviewer 2: You need to actually indent the whole block, not just the code inside the block. Here's an example of what it should look like:

<img width="300" alt="image" src="https://github.com/arrocke/gloss-translation/assets/17209828/40e9c2e5-5d85-4944-bd6d-f496b6c78c00">

Reviewer 2: This applies to all the code blocks in this file. Do you have a markdown preview in your VS code? If you preview the markdown, it will make it obvious where your indentation needs adjusted.

To see which boxes are installed run:

```bash
vagrant box list
```

8. Copy file: `Vagrantfile` from host `gloss-translation` repository `.vagrant/` folder to your vagrant project folder.
9. Read the `NOTES:` section contained in file `Vagrantfile`.
10. in the `Vagrantfile`, edit `HOST_FOLDER`

```ruby
 HOST_FOLDER = "../gloss-translation"
```

> Replace `../gloss-translation` with the location of your repository root folder.

Reviewer 2: ```suggestion
    Replace `../gloss-translation` with the location of your repository root folder.
```

Author: `Replace...` line is now integrated into the line before the code block.

   1. Launch the `Local Security Policy` app from the `start menu` (or `Win+R`, then type `secpol.msc`)
   1. Navigate to `Local Policies` -> `User Rights Assignment`.
   1. Open the `Create symbolic links` property. By default it has only `Administrators` listed.
   1. Click `Add User or Group...` button and type `Authenticated Users` in the `Enter the object names to select` field and then press the `OK` button. This will grant `symbolic link` priviledge to all users who log on with credetials.
   1. Close any open windows by clicking `OKs` and and then `log off` and `log on` to activate.
   1. Symbolic links should now work in VirtualBox shared folders without administrator priviledges.

## Setup Steps

1. [Download](https://developer.hashicorp.com/vagrant/downloads?ajs_aid=ba208f95-d5a1-457d-abeb-49d458b95eec&product_intent=vagrant) Vagrant software.
1. Reference these [installation](https://developer.hashicorp.com/vagrant/docs/installation) instructions.
1. Verify installation by running `vagrant -v` in a terminal session.
1. Install [VirtualBox](https://www.virtualbox.org/).
1. If you have not already done so, fork repo to your local machine.

> 1. Vagrant uses a special configuration file called `Vagrantfile` that contains all the information required for the creation of a Virtual Machine (guest) on your workstation (host) . It is located n your repo root directory.

Reviewer 2: ```suggestion
1. Vagrant uses a special configuration file called `Vagrantfile` that contains all the information required for the creation of a Virtual Machine (guest) on your workstation (host). It is located in your repo root directory.
```

You missed a few things in the last correction.

Reviewer 2: You can use the "Commit suggestion" button to avoid missing anything.

Author: Never noticed that button.  Thanks

Author: This is a response to your comment below:

How long does it usually take to boot the VM? This is at the bottom of the output of `vagrant up`.

```text
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
```

I have experienced this issue a number of times as I was working this PR. It appears to be a bug in the relationship between vagrant and virtualbox.  At this point in the process, the VM is up and running and vagrant is trying to establish a SSH session to it. However, it hangs indefinitely at this point and vagrant eventually times out. There is a vagrantfile line `config.vm.boot_timeout` were you can increase the timeout value, however, this would not solve the problem as you would just have to wait long before it times out.  

I don't know why this happens. It happens inconsistently, which makes it hard to debug.

If you find that vagrant is stuck at this point, you can free it up by performing the following steps:
1. Open `Oracle VM VirtualBox Manager`
2. Click your VM machine in the left panel. The name starts with the directory name of the root of your repository. In our case it would be `gloss-translation_....`.
3. If the VM is running, you will see a green right arrow called `Show`. It opens a new panel where you can monitor the boot process of your VM.  For some unknown reason, when I do this, it frees up the vagrant session and then continue booting up to completion.
4. Once the vagrant session is up, you can close down the panel, without shutting down the VM by clicking menu `Machine/Detach GUI`.

This was one of the reasons why I thought we should let developers know that this tool exists.  It does have a lot of useful features in it.

Let me know if you want to add some comment about this issue in the `Caveat` section.

Reviewer 2: These steps don't work for me because I don't seem to have VirtualBox manager installed.

Reviewer 2: I'm guessing this will also be true for anyone who has virtualbox installed for them by vagrant. This is probably because we don't need the full suite of virtualbox tools installed to use vagrant.

Reviewer 2: I think we can leave this be for now. Looking at StackOverflow, it seems like vagrant integration with virtualbox has some common issues. I look forward to using the docker configuration instead. I think it might be the better option since we already use docker in this project for other things.

1. To start the creation of your virtual environment, run this command in the repo root directory. (This takes a while so be patient.)

   ```bash
   vagrant up
   ```

> 1. Once the `vagrant up` command completes, you can logon into your virutal machine using this command. (password: `vagrant`)

Reviewer 2: ```suggestion
1. Once the `vagrant up` command completes, you can logon into your virtual machine using this command. (password: `vagrant`)
```

I would recommend installing the "Code Spell Checker" extension in VS Code, so it is easier to catch these issues.

Author: done.
Thanks for recommending the spell checker.  Wanted one, but did not know that it existed.

   | `vagrant up` --provision \| tee provision.log | Runs vagrant up, forces provisioning and logs all output to a file                                                                   |
   | `VAGRANT_LOG=info`                            | Use the environement variable VAGRANT_LOG to set verbosity. Enter this in terminal session before you run the `vagrant up` command   |

1. For help on vagrant commands, at a terminal session enter:

   ```bash
   vagrant --help
   ```

## Post-Installation Notes

Accessing VM resources from Host.

### Accessing database via pgAdmin

1. Start pgAdmin
1. Open the `Register - Server` dialog
1. `General` tab:
   - Name: _Enter name of your choice_
1. `Connection` tab

   | Setting               | Value       |
   | --------------------- | ----------- |
   | Host name:            | `127.0.0.1` |
   | Port:                 | `5432`      |
   | Maintenance database: | `postgres`  |
   | Username:             | `vagrant`   |
   | Password:             | `vagrant`   |
   | Save password?:       | `Turn on`   |

### Accessing API app

> Open a web brower to URL: <http://localhost:4300>

Reviewer 2: ```suggestion
Open a web browser to URL: <http://localhost:4300>
```

Author: done.

### Accessing Web app

> Open a web brower to URL: <http://localhost:4200>

Reviewer 2: ```suggestion
Open a web browser to URL: <http://localhost:4200>
```

Author: done

   1. Launch the `Local Security Policy` app from the `start menu` (or `Win+R`, then type `secpol.msc`)
   1. Navigate to `Local Policies` -> `User Rights Assignment`.
   1. Open the `Create symbolic links` property. By default it has only `Administrators` listed.
   1. Click `Add User or Group...` button and type `Authenticated Users` in the `Enter the object names to select` field and then press the `OK` button. This will grant `symbolic link` priviledge to all users who log on with credetials.
   1. Close any open windows by clicking `OKs` and and then `log off` and `log on` to activate.
   1. Symbolic links should now work in VirtualBox shared folders without administrator priviledges.

## Setup Steps

1. [Download](https://developer.hashicorp.com/vagrant/downloads?ajs_aid=ba208f95-d5a1-457d-abeb-49d458b95eec&product_intent=vagrant) Vagrant software.
1. Reference these [installation](https://developer.hashicorp.com/vagrant/docs/installation) instructions.
1. Verify installation by running `vagrant -v` in a terminal session.
1. Install [VirtualBox](https://www.virtualbox.org/).
1. If you have not already done so, fork repo to your local machine.

> 1. Vagrant uses a special configuration file called: `Vagrantfile` that contains all the information required for the creation of a Virtual Machine (guest) on your workstation (host) . It is located n your repo root directory.

Reviewer 2: ```suggestion
1. Vagrant uses a special configuration file called `Vagrantfile` that contains all the information required for the creation of a Virtual Machine (guest) on your workstation (host). It is located in the root directory of the repo.
```

Author: Done.

1. To start the creation of your virtual environment, run this command in the repo root directory. (This takes a while so be patient.)

   ```bash
   vagrant up
   ```

1. Once the `vagrant up` command completes, you can logon into your virutal machine using this command. (password: `vagrant`)

   ```bash
   vagrant ssh
   ```

> 1. Here is a list of useful Vagrant commands. Be sure that you are in the same directory as the Vagran­tfile when running these commands!

Reviewer 2: Is it necessary to have this list, when we can just use `vagrant --help` to get the available commands? Having this list in our docs requires us to keep it in sync with vagrant's docs. If they update or remove a command, our docs will be out of date.

Reviewer 2: Some of the items like `VAGRANT_LOG=info` would be useful for our docs, because they are not listed in `vagrant --help`

Reviewer 2: Maybe we should include the very basic commands like `vagrant up`, `vagrant halt`, and `vagrant ssh` (basically only the most basic commands which we'll need for day-to-day usage). Then we can refer developers to `vagrant --help` if they want to go deeper.

@arrocke What are your thoughts? Am I way off?

Reviewer: Yep, this documentation should be enough for a developer to get started, and refer to the vagrant docs for more detail. I think we can assume that devs that want to use vagrant, will do their own research or will already be familiar

Author: So just to be clear. Do you want me to remove point 9.  Here is a list of usefule Vagrant commands....    but leave point 10?

Reviewer: I'm ok removing both. Devs can do their own research if they want to do more than the basic commands to get everything running

Reviewer 2: @ilearner777 Go ahead and remove points 9 and 10. You should replace them with a point that describes the basic commands, just "enough for a developer to get started" like Adrian said.

Author: I have replaced point 9 with the following:

9. Here is a list of common Vagrant commands. Be sure that you are in the same directory as the Vagrantfile when running these commands!

   |Command|Description|
   |-------|-----------|
   |`vagrant up`|starts vagrant environment (also provisions only on the FIRST vagrant up) Equivalent to pressing the power buttons on your servers.|
   |`vagrant ssh`|connects to machine via SSH|
   |`vagrant suspend`|Suspends a virtual machine (remembers state). Starts faster.|
   |`vagrant halt`|stops the vagrant machine. Similar to shutdown.|
   |`vagrant status`|outputs status of the vagrant machine|

   | `vagrant box remove` [virtualbox name]        | delete a box                                                                                                                         |
   | `vagrant global­-status`                      | outputs status of all vagrant machines                                                                                               |
   | `vagrant global­-status` --prune              | same as above, but prunes invalid entries                                                                                            |
   | `vagrant up` --provision \| tee provision.log | Runs vagrant up, forces provisioning and logs all output to a file                                                                   |
   | `VAGRANT_LOG=info`                            | Use the environement variable VAGRANT_LOG to set verbosity. Enter this in terminal session before you run the `vagrant up` command   |

1. For help on vagrant commands, at a terminal session enter:

   ```bash

>    vagrant -help

Reviewer 2: ```suggestion
   vagrant --help
```

Author: Done.

   ```

> 1. Once your VM environment is created, as an alternative to the vagrant commands listed above, you can use the `Oracle VM VirtualBox` application to manage the starting, stopping, and configuration of your VM. The `Help` menu provides detailed instructions on its use.

Reviewer 2: Will using Virtualbox directly break the vagrant setup?

It seems unnecessary to mention this, when we can just use vagrant to manage our VMs in a consistent manner, regardless of provider.

Reviewer: Opening virtualbox shouldn't break vagrant, sometimes that is useful for debugging what is going on. But I agree, this shouldn't be necessary in our docs

Author: I remove point 11.

1. Verify installation by running `vagrant -v` in a terminal session.
1. Install [VirtualBox](https://www.virtualbox.org/).
1. If you have not already done so, fork repo to your local machine.
1. Next you will need to install a [box](https://developer.hashicorp.com/vagrant/tutorials/getting-started/getting-started-boxes). The `Vagrantfile` in this repository uses box: `generic/ubuntu2204`. So to install this box, run in a terminal window:

```bash

>     vagrant box add generic/ubuntu2204

Reviewer 2: You mentioned virtualbox before, but it would be good to mention it again here so people don't get confused when vagrant asks them which provider to use.

Author: I believe this only occurs when you use command `vagrant init`.  When I retested  on 2023-10-28 following my latest instructions, vagrant did not ask me which provider to use.
